go-controller/pkg/ovn/egressip.go

```diff
@@ -268,63 +268,87 @@ func (oc *Controller) syncStaleEgressReroutePolicy(egressIPToPodIPCache map[stri
+		podIPCache, exists := egressIPToPodIPCache[egressIPName]
+		return !exists || !podIPCache.Has(parsedLogicalIP)
+	}).List(&nats)
+
 	if err != nil {
-		klog.Errorf("Unable to sync egress IPs, unable to find NAT IDs, stderr: %s, err: %v", stderr, err)
+		klog.Errorf("Unable to sync egress IPs, unable to find NAT IDs, err: %v", err)
 		return
 	}
-	for _, natItem := range strings.Split(natItems, "\n") {
-		if natItem == "" {
-			continue
-		}
-		natFields := strings.Split(natItem, ",")
 
-		UUID := natFields[0]
-		externalID := natFields[1]
-		logicalIP := natFields[2]
+	if len(nats) == 0 {
+		// No stale nat entries to deal with: noop.
+		return
+	}
 
-		parsedLogicalIP := net.ParseIP(logicalIP).String()
-		if !strings.Contains(externalID, "name") {
-			continue
-		}
-		egressIPName := strings.Split(externalID, "=")[1]
+	natUUIDs := make([]string, 0, len(nats))
+	for _, nat := range nats {
+		natUUIDs = append(natUUIDs, nat.UUID)
+	}
+	natUUIDsSet := sets.NewString(natUUIDs...)
 
-		podIPCache, exists := egressIPToPodIPCache[egressIPName]
-		if !exists || !podIPCache.Has(parsedLogicalIP) {
-			logicalRouters, stderr, err := util.RunOVNNbctl(
-				"--format=csv",
-				"--data=bare",
-				"--no-heading",
-				"--columns=name",
-				"find",
-				"logical_router",
-				fmt.Sprintf("nat{>=}%s", UUID),
-			)
-			if err != nil {
-				klog.Errorf("Unable to find logical_router associated with stale NAT ID: %s, stderr: %s, err: %v", UUID, stderr, err)
-			}
-			for _, logicalRouter := range strings.Split(logicalRouters, "\n") {
-				if logicalRouter != "" {
-					_, stderr, err = util.RunOVNNbctl(
-						"remove",
-						"logical_router",
-						logicalRouter,
-						"nat",
-						UUID,
-					)
-					if err != nil {
-						klog.Errorf("Unable to delete stale NAT ID: %s associated with logical router: %s, stderr: %s, err: %v", UUID, logicalRouter, stderr, err)
-					}
-				}
+	// At this point, we have a list of NATs that should be removed. In order to
+	// to that, iterate through the routers and identify which ones have these nat(s).
+	routers := []nbdb.LogicalRouter{}
+	err = oc.nbClient.WhereCache(func(item *nbdb.LogicalRouter) bool {
+		for _, rtrNatUUID := range item.Nat {
+			if natUUIDsSet.Has(rtrNatUUID) {
+				return true
 			}
 		}
+		return false
+	}).List(&routers)
+
+	if err != nil {
+		klog.Errorf("Unable to sync egress IPs, unable to find routers, err: %v", err)
+		return
+	}
+
+	ops := []ovsdb.Operation{}
+	for _, natUUID := range natUUIDs {
+		op, err := oc.nbClient.Where(
+			&nbdb.NAT{
+				UUID: natUUID,
+			}).Delete()
+
+		if err != nil {
+			klog.Errorf("Error deleting stale NAT %s: %v", natUUID, err)
+			return
+		}
+		ops = append(ops, op...)
+	}
+
+	// Removing stale NATS: since mutate delete of values that are not present is okay,
+	// the current implementation keeps stale uuids in a 'simple' single slice and
+	// uses that for all routers that have a subset with any of them.
+	for _, router := range routers {
```

Comment:
Also here: you have a `DeleteNatsFromRouter`, why not use that?

Reply:
UPDATE: I will make findRouter faster / cheaper and then leverage DeleteNatsFromRouter.

Reply:
done.